File: btypes_big_integer/src/main/cpp/BBoolean.cpp
#include <iostream>
#include <string>
#include "BObject.cpp"

using namespace std;


#ifndef BBOOLEAN_H
#define BBOOLEAN_H

class BBoolean : public BObject {
	private:
	    bool value;

    public:
        typedef BBoolean current_type;
        typedef void value_type;
        typedef void left_type;
        typedef void right_type;

    	BBoolean(bool val) {
    		value = val;
    	}

    	BBoolean(){};

    	BBoolean(const BBoolean& val) {
    	    value = val.value;
    	}

    	static string toString(bool b) {
    		return b ? "true" : "false";
    	}

	    bool booleanValue() const {
		    return value;
	    }

        BBoolean _or(const BBoolean& other) const {
            return value || other.value;
        }

        BBoolean _or(bool other) const {
            return value || other;
        }

        BBoolean _xor(const BBoolean& other) const {
            return value ^ other.value;
        }

        BBoolean _xor(bool other) const {
            return value ^ other;
        }

        BBoolean _and(const BBoolean& other) const {
            return value && other.value;
        }

        BBoolean _and(bool other) const {
            return value && other;
        }

        BBoolean _not() const {
            return !value;
        }

        BBoolean implies(const BBoolean& other) const {
            return !value || other.value;
        }

        BBoolean implies(bool other) const {
            return !value || other;
        }

        BBoolean equivalent(bool other) const {
            return value == other;
        }

        BBoolean equivalent(const BBoolean& other) const {
            return value == other.value;
        }

        BBoolean equal(const BBoolean& other) const {
            return value == other.value;
        }

        BBoolean unequal(const BBoolean& other) const {
            return value != other.value;
        }

        friend bool operator !=(const BBoolean& o1, const BBoolean& o2) {
            return o1.value != o2.value;
        }

        friend bool operator ==(const BBoolean& o1, const BBoolean& o2) {
            return o1.value == o2.value;
        }

        bool operator <(const BBoolean& other) {
            return value != other.value;
        }

        void operator =(const BBoolean& other) {
            value = other.value;
        }

        int hashCode() const {
            return value == true ? 1 : 0;
        }

        friend std::ostream& operator<<(std::ostream &strm, const BBoolean &b) {
          return strm << toString(b.value);
        }
};
#endif

File: btypes_primitives/src/main/python_magicstack_immutable/btypes/BBoolean.py
class BBoolean:
    def __init__(self, value):
        if type(value) is bool:
            self.__value = value
        elif type(value) is str:
            self.__value = str(value).lower() == "true"
        elif type(value) is BBoolean:
            self.__value = value.__value

    def __bool__(self) -> 'bool':
        return self.__value

    def __and__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value and other.__value)

    def __or__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value or other.__value)

    def __rand__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value and other.__value)

    def __str__(self) -> 'str':
        return str(self.__value).lower()

    def __ror__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value or other.__value)

    def __rxor__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value ^ other.__value)

    def __xor__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value ^ other.__value)

    def __eq__(self, other: 'BBoolean') -> 'BBoolean':
        if not isinstance(other, BBoolean):
            return BBoolean(False)
        return BBoolean(self.__value == other.__value)

    def __ne__(self, other: 'BBoolean') -> 'BBoolean':
        return BBoolean(self.__value != other.__value)

    def _and(self, other: 'BBoolean') -> 'BBoolean':
        return self.__and__(other)

    def _or(self, other: 'BBoolean') -> 'BBoolean':
        return self.__or__(other)

    def _not(self) -> 'BBoolean':
        return BBoolean(not self.__value)

    def implies(self, other: 'BBoolean') -> 'BBoolean':
        return self._not()._or(other)

    def equivalent(self, other: 'BBoolean') -> 'BBoolean':
        return self.implies(other)._and(other.implies(self))

    def equal(self, other: 'BBoolean') -> 'BBoolean':
        return self.__eq__(other)

    def unequal(self, other: 'BBoolean') -> 'BBoolean':
        return self.__ne__(other)

    def booleanValue(self) -> 'bool':
        return self.__value

    def __hash__(self):
        return hash(self.__value)
